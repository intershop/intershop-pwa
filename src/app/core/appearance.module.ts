import { NgModule } from '@angular/core';
import { NgbPopoverConfig } from '@ng-bootstrap/ng-bootstrap';
import { ToastrModule } from 'ngx-toastr';

import { IconModule } from './icon.module';
import { ThemeService } from './utils/theme/theme.service';

@NgModule({
  imports: [
    IconModule,
    ToastrModule.forRoot({
      closeButton: true,
      timeOut: 3000,
      positionClass: 'toast-top-full-width', // toast-top-center
      preventDuplicates: true,
    }),
  ],
})
export class AppearanceModule {
  constructor(popoverConfig: NgbPopoverConfig, themeService: ThemeService) {
    popoverConfig.placement = 'top';
    popoverConfig.triggers = 'hover';
    popoverConfig.container = 'body';

    themeService.init();
  }
}
