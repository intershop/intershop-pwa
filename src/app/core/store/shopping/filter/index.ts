// API to access ngrx filter state
export * from './filter.actions';
export * from './filter.selectors';
