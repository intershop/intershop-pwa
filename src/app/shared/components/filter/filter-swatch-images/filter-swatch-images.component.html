<ul class="filter-list clearfix">
  <ng-container *ngFor="let facet of filterElement.facets">
    <li class="filter-item filter-color" [ngClass]="{ 'filter-selected': facet.selected }">
      <a
        class="filter-swatch"
        [title]="facet.displayName + ' (' + facet.count + ')'"
        (click)="filter(facet)"
        [attr.data-testing-id]="'filter-link-' + (facet.displayName | sanitize)"
      >
        <span
          [style.background-image]="getBackgroundImage(facet)"
          [style.background-color]="getBackgroundColor(facet)"
          [style.border]="facet.displayName.toLowerCase() === 'white' ? '1px solid #ddd' : null"
        ></span>
      </a>
    </li>
  </ng-container>
</ul>
