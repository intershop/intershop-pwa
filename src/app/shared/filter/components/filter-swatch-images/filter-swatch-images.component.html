<ul class="filter-list clearfix">
  <ng-container *ngFor="let facet of filterElement.facets">
    <li class="filter-item filter-color" [ngClass]="{ 'filter-selected': facet.selected }">
      <a
        class="filter-swatch"
        title="facet.name"
        [title]="facet.name"
        (click)="filter(facet)"
        [attr.data-testing-id]="'filter-link-' + (facet.name | sanitize)"
      >
        <!--- TODO: use color code from REST call as background color-->
        <span [style.background-color]="facet.name.toLowerCase()"></span>
      </a>
    </li>
  </ng-container>
</ul>
