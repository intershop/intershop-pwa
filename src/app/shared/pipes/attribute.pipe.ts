import { formatDate, formatNumber } from '@angular/common';
import { Pipe, PipeTransform } from '@angular/core';
import { TranslateService } from '@ngx-translate/core';
import { Attribute } from '../../models/attribute/attribute.model';
import { Price } from '../../models/price/price.model';
import { formatPrice } from './price.pipe';

@Pipe({ name: 'ishAttribute', pure: false })
export class AttributeToStringPipe implements PipeTransform {
  constructor(private translateService: TranslateService) {}

  private toDate(val): string {
    return formatDate(val, 'shortDate', this.translateService.currentLang);
  }

  private toDecimal(val): string {
    return formatNumber(val, this.translateService.currentLang);
  }

  private toCurrency(price: Price): string {
    return formatPrice(price, this.translateService.currentLang);
  }

  transform(data: Attribute, valuesSeparator: string = ', '): string {
    if (!this.translateService.currentLang) {
      return 'undefined';
    }

    switch (data.type) {
      case 'String':
        return data.value;
      case 'Integer':
      case 'Double':
      case 'Long':
      case 'BigDecimal':
        return this.toDecimal(data.value);
      case 'Boolean':
        return data.value.toString();
      case 'Date':
        // TODO: ISREST-222 - waiting for adaption of REST API response to return Date values not as 'String' so they can be handled accordingly
        return this.toDate(data.value);
      case 'MultipleInteger':
      case 'MultipleDouble':
      case 'MultipleLong':
      case 'MultipleBigDecimal':
        return data.value.map(v => this.toDecimal(v)).join(valuesSeparator);
      case 'MultipleString':
      case 'MultipleBoolean':
        return data.value.join(valuesSeparator);
      case 'MultipleDate':
        return data.value.map(v => this.toDate(v)).join(valuesSeparator);
      case 'ResourceAttribute':
        switch (data.value.type) {
          case 'Quantity':
            return `${this.toDecimal(data.value.value)}\xA0${data.value.unit}`;
          case 'Money':
            return this.toCurrency(data.value as Price);
          default:
            return data.value.toString();
        }
      default:
        return data.value.toString();
    }
  }
}
