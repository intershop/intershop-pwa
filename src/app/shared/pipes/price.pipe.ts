import { formatCurrency, getCurrencySymbol } from '@angular/common';
import { Pipe, PipeTransform } from '@angular/core';
import { TranslateService } from '@ngx-translate/core';

import { Price } from 'ish-core/models/price/price.model';

export function formatPrice(price: Price, lang: string): string {
  const symbol = getCurrencySymbol(price.currencyMnemonic, 'wide', lang);
  return formatCurrency(price.value, lang, symbol);
}

@Pipe({ name: 'ishPrice', pure: false })
export class PricePipe implements PipeTransform {
  constructor(private translateService: TranslateService) {}

  transform(data: Price): string {
    if (!data || !this.translateService.currentLang) {
      return 'undefined';
    }
    switch (data.type) {
      case 'Money':
      case 'ProductPrice':
        return formatPrice(data, this.translateService.currentLang);
      default:
        return data.toString();
    }
  }
}
