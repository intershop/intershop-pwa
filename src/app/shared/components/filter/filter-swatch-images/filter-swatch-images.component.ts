import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output } from '@angular/core';

import { Facet } from 'ish-core/models/facet/facet.model';
import { Filter } from 'ish-core/models/filter/filter.model';
import { URLFormParams } from 'ish-core/utils/url-form-params';

/**
 * The Filter Swatch Images Component displays filter group for colors. The facets of the filter group are presented as swatch images.
 *
 * @example
 * <ish-filter-dropdown
 *               [filterElement]="element"
 *               (applyFilter)="applyFilter($event)"
 * </ish-filter-dropdown>
 */
@Component({
  selector: 'ish-filter-swatch-images',
  templateUrl: './filter-swatch-images.component.html',
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class FilterSwatchImagesComponent {
  /**
   * The filter group.
   */
  @Input() filterElement: Filter;
  @Output() applyFilter: EventEmitter<{ searchParameter: URLFormParams }> = new EventEmitter();

  /**
   * Applies a facet of the filter group and shows the new filtered result.
   */
  filter(facet: Facet) {
    this.applyFilter.emit({ searchParameter: facet.searchParameter });
  }

  getBackgroundColor(facet: Facet) {
    return facet.mappedType === 'colorcode' ? facet.mappedValue : undefined;
  }

  getBackgroundImage(facet: Facet) {
    return facet.mappedType === 'image' ? facet.mappedValue : undefined;
  }
}
