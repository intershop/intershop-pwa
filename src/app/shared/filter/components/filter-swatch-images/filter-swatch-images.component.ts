import { ChangeDetectionStrategy, Component, EventEmitter, Input, OnInit, Output } from '@angular/core';
import { FormControl, FormGroup } from '@angular/forms';
import { Subject } from 'rxjs';
import { takeUntil } from 'rxjs/operators';

import { Facet } from 'ish-core/models/facet/facet.model';
import { Filter } from 'ish-core/models/filter/filter.model';

/**
 * The Filter Swatch Images Component displays filter group for colors. The facets of the filter group are presented as swatch images.
 *
 * @example
 * <ish-filter-dropdown
 *               [filterElement]="element"
 *               (applyFilter)="applyFilter($event)"
 * </ish-filter-dropdown>
 */
@Component({
  selector: 'ish-filter-swatch-images',
  templateUrl: './filter-swatch-images.component.html',
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class FilterSwatchImagesComponent implements OnInit {
  /**
   * The filter group.
   */
  @Input() filterElement: Filter;
  @Output() applyFilter: EventEmitter<{ searchParameter: string }> = new EventEmitter();

  filterForm: FormGroup;

  private destroy$ = new Subject();

  ngOnInit() {
    this.filterForm = new FormGroup({
      filterDropdown: new FormControl(''),
    });

    this.filterForm
      .get('filterDropdown')
      .valueChanges.pipe(takeUntil(this.destroy$))
      .subscribe(filterDropdown => this.filter(filterDropdown));
  }

  /**
   * Applies a facet of the filter group and shows the new filtered result.
   */
  filter(facet: Facet) {
    this.applyFilter.emit({ searchParameter: facet.searchParameter });
  }
}
