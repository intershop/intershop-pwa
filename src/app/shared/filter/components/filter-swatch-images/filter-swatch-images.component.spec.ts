import { ComponentFixture, TestBed, async, fakeAsync, tick } from '@angular/core/testing';
import { NgbCollapseModule } from '@ng-bootstrap/ng-bootstrap';

import { IconModule } from 'ish-core/icon.module';
import { Filter } from 'ish-core/models/filter/filter.model';
import { PipesModule } from 'ish-core/pipes.module';

import { FilterSwatchImagesComponent } from './filter-swatch-images.component';

describe('Filter Swatch Images Component', () => {
  let component: FilterSwatchImagesComponent;
  let fixture: ComponentFixture<FilterSwatchImagesComponent>;
  let element: HTMLElement;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      imports: [IconModule, NgbCollapseModule, PipesModule],
      declarations: [FilterSwatchImagesComponent],
    }).compileComponents();
  }));

  beforeEach(() => {
    const filterElement = {
      name: 'Color',
      facets: [
        { name: 'Black', count: 4, displayName: 'Black' },
        { name: 'Red', count: 5, displayName: 'Red', selected: true },
      ],
    } as Filter;
    fixture = TestBed.createComponent(FilterSwatchImagesComponent);
    component = fixture.componentInstance;
    element = fixture.nativeElement;
    component.filterElement = filterElement;
    fixture.detectChanges();
  });

  it('should be created', () => {
    expect(component).toBeTruthy();
    expect(element).toBeTruthy();
    expect(() => fixture.detectChanges()).not.toThrow();
    expect(element).toMatchSnapshot();
  });

  it('should toggle all swatch images when filter group header is clicked', fakeAsync(() => {
    fixture.detectChanges();
    const filterGroupHead = fixture.nativeElement.querySelectorAll('h3')[0];
    filterGroupHead.click();
    tick(500);
    fixture.detectChanges();

    const hiddenFilterFacet = element.getElementsByTagName('ul')[0];
    expect(hiddenFilterFacet.className).not.toContain('show');
  }));
});
