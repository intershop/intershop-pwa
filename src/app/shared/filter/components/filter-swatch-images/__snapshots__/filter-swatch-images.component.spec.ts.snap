// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter Swatch Images Component should be created 1`] = `
<div class="filter-group">
  <h3 aria-controls="collapseFilter" aria-expanded="true">
    Color <fa-icon class="float-right" ng-reflect-icon-prop="fas,angle-down"></fa-icon>
  </h3>
  <ul class="filter-list clearfix" id="collapseFilter" ng-reflect-collapsed="false">
    <li class="filter-item filter-colour">
      <a class="filter-swatch" title="Black" data-testing-id="filter-link-Black"
        ><span style="background-color: black;"></span
      ></a>
    </li>
    <li class="filter-item filter-colour filter-selected">
      <a class="filter-swatch" title="Red" data-testing-id="filter-link-Red"
        ><span style="background-color: red;"></span
      ></a>
    </li>
  </ul>
</div>
`;
